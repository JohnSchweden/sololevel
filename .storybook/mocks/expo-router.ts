// Mock Expo Router hooks for Storybook
import React from 'react'

export const Link = ({ href, children, asChild, ...props }: any) => {
  if (asChild) {
    return children
  }
  return React.createElement('a', { href, ...props }, children)
}

export const useRouter = () => ({
  push: (_href: string) => {},
  replace: (_href: string) => {},
  back: () => {},
  canGoBack: () => false,
})

export const useLocalSearchParams = () => ({ id: 'mock-id' })

export const Stack = {
  Screen: ({ options, children }: any) => children || null,
}
