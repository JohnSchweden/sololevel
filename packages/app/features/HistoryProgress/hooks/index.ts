export { useHistoryQuery } from './useHistoryQuery'
export type { VideoItem } from './useHistoryQuery'
