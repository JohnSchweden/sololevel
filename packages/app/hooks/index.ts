export { useConfirmDialog } from './useConfirmDialog'
export type { ConfirmDialogState } from './useConfirmDialog'
export { useStatusBar } from './useStatusBar'
export { useAnimationCompletion } from './useAnimationCompletion'
export type {
  UseAnimationCompletionConfig,
  UseAnimationCompletionReturn,
} from './useAnimationCompletion'
export { useSmoothnessTracking } from './useSmoothnessTracking'
export type {
  UseSmoothnessTrackingConfig,
  UseSmoothnessTrackingReturn,
  SmoothnessMetrics,
} from './useSmoothnessTracking'
export { useFrameDropDetection } from './useFrameDropDetection'
export type {
  UseFrameDropDetectionConfig,
  UseFrameDropDetectionReturn,
  FrameDropMetrics,
} from './useFrameDropDetection'
