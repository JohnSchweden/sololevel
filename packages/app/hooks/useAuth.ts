import { type AuthResult, type SignInData, authClient } from '@my/api'
import { log } from '@my/logging'
import { useCallback, useEffect } from 'react'
import { useAuthStore } from '../stores/auth'

/**
 * Auth hook that provides authentication state and actions
 * Uses the new typed authClient and integrates with Zustand store
 */
export function useAuth() {
  const { user, session, loading, initialized, setAuth, setLoading, setInitialized } =
    useAuthStore()

  // Log deduplication ref - must be called unconditionally at hook level
  // Removed unused logOnce helper

  /**
   * Initialize auth state and set up listeners
   */
  const initialize = useCallback(async () => {
    if (initialized) return

    log.info('useAuth', 'Initializing auth state')
    setLoading(true)

    try {
      // Get initial session using our typed auth client
      const sessionResult = await authClient.getSession()

      if (sessionResult.success && sessionResult.data) {
        setAuth(sessionResult.data.user, sessionResult.data)
        log.info('useAuth', 'Found existing session', {
          userId: sessionResult.data.user.id,
        })
      } else {
        setAuth(null, null)
        log.info('useAuth', 'No existing session found')
      }

      // Set up auth state listener
      const unsubscribe = authClient.onAuthStateChange((session) => {
        if (__DEV__) {
          log.debug('useAuth', 'Auth state changed', {
            hasSession: !!session,
            userId: session?.user?.id,
          })
        }
        setAuth(session?.user ?? null, session)
      })

      setInitialized(true)
      setLoading(false)

      // Return cleanup function
      return unsubscribe
    } catch (error) {
      log.error('useAuth', 'Failed to initialize auth', { error })
      setAuth(null, null)
      setInitialized(true)
      setLoading(false)
      return undefined
    }
  }, [initialized, setAuth, setLoading, setInitialized])

  /**
   * Sign in with email and password
   */
  const signIn = useCallback(
    async (email: string, password: string): Promise<AuthResult<SignInData>> => {
      log.info('useAuth', 'Attempting sign in', {
        email: email.replace(/(.{2}).*(@.*)/, '$1***$2'), // Mask email for privacy
      })

      setLoading(true)

      const result = await authClient.signInWithPassword(email, password)

      if (result.success) {
        // Auth state will be updated via the listener
        log.info('useAuth', 'Sign in successful', {
          userId: result.data.user.id,
        })
      } else {
        log.warn('useAuth', 'Sign in failed', {
          code: result.error.code,
          message: result.error.message,
        })
      }

      setLoading(false)
      return result
    },
    [setLoading]
  )

  /**
   * Sign out current user
   */
  const signOut = useCallback(async (): Promise<AuthResult<void>> => {
    log.info('useAuth', 'Attempting sign out')
    setLoading(true)

    const result = await authClient.signOut()

    if (result.success) {
      // Auth state will be updated via the listener
      log.info('useAuth', 'Sign out successful')
    } else {
      log.warn('useAuth', 'Sign out failed', {
        code: result.error.code,
        message: result.error.message,
      })
    }

    setLoading(false)
    return result
  }, [setLoading])

  /**
   * Get current user ID (convenience method)
   */
  const userId = user?.id ?? null

  /**
   * Check if user is authenticated
   */
  const isAuthenticated = !!user && !!session

  // Initialize on mount
  useEffect(() => {
    let cleanup: (() => void) | undefined

    initialize().then((unsubscribe) => {
      cleanup = unsubscribe
    })

    return () => {
      cleanup?.()
    }
  }, [initialize])

  return {
    // State
    user,
    session,
    loading,
    initialized,
    userId,
    isAuthenticated,

    // Actions
    signIn,
    signOut,
    initialize,
  }
}
