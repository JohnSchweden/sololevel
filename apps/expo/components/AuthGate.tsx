import { useAuth } from '@my/app/hooks/useAuth'
import { log } from '@my/logging'
import { H3, Spinner, YStack } from '@my/ui'
import { usePathname, useRouter } from 'expo-router'
import { useEffect, useRef } from 'react'

export interface AuthGateProps {
  children: React.ReactNode
  redirectTo?: string
  fallback?: React.ReactNode
}

/**
 * AuthGate - Route protection component for Expo Router
 *
 * Protects child routes by:
 * - Showing loading state while auth is initializing
 * - Redirecting unauthenticated users to sign-in
 * - Rendering children only when user is authenticated
 * - Preserving intended destination for post-auth redirect
 */
export function AuthGate({ children, redirectTo = '/auth/sign-in', fallback }: AuthGateProps) {
  const { isAuthenticated, loading, initialized } = useAuth()
  const router = useRouter()
  const pathname = usePathname()

  // Check if current route is an auth route (should be accessible without authentication)
  const isAuthRoute = pathname.startsWith('/auth')

  if (typeof __DEV__ !== 'undefined' && __DEV__) {
    log.debug('AuthGate', 'Render', { isAuthenticated, loading, initialized })
  }

  // Log deduplication ref - must be called unconditionally at component level
  const lastLoggedKeyRef = useRef<string>('')

  // Log deduplication helper to prevent spam of the same transition
  const logOnce = (key: string, fn: () => void) => {
    if (lastLoggedKeyRef.current !== key) {
      lastLoggedKeyRef.current = key
      fn()
    }
  }

  useEffect(() => {
    // Only redirect after auth is initialized and user is not authenticated
    // Skip redirect if we're already on an auth route
    if (initialized && !loading && !isAuthenticated && !isAuthRoute) {
      log.info('AuthGate', 'User not authenticated, redirecting to sign-in', {
        redirectTo,
      })

      // Use replace to avoid adding to navigation stack
      router.replace(redirectTo as any)
    }
  }, [initialized, loading, isAuthenticated, redirectTo, router, isAuthRoute])

  // Allow auth routes to pass through without authentication check
  if (isAuthRoute) {
    return <>{children}</>
  }

  // Show loading state while auth is initializing
  if (!initialized || loading) {
    if (typeof __DEV__ !== 'undefined' && __DEV__) {
      log.debug('AuthGate', 'Auth initializing, showing loading state', {
        initialized,
        loading,
      })
    }

    if (fallback) {
      return <>{fallback}</>
    }

    return (
      <YStack
        flex={1}
        justifyContent="center"
        alignItems="center"
        gap="$4"
        backgroundColor="$background"
      >
        <Spinner size="large" />
        <H3 color="$color10">Loading...</H3>
      </YStack>
    )
  }

  // Don't render children if not authenticated (redirect will happen)
  if (!isAuthenticated) {
    logOnce('not-authenticated', () =>
      log.info('AuthGate', 'User not authenticated, not rendering children')
    )
    return null
  }

  // User is authenticated, render protected content
  logOnce('authenticated', () =>
    log.info('AuthGate', 'User authenticated, rendering protected content')
  )
  return <>{children}</>
}
