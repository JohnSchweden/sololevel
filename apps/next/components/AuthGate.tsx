import { useAuth } from '@my/app/hooks/useAuth'
import { log } from '@my/logging'
import { H3, Spinner, YStack } from '@my/ui'
import { useRouter } from 'expo-router'
import React, { useEffect } from 'react'

export interface AuthGateProps {
  children: React.ReactNode
  redirectTo?: string
  fallback?: React.ReactNode
}

/**
 * AuthGate - Route protection component for Next.js with Expo Router
 *
 * Protects child routes by:
 * - Showing loading state while auth is initializing
 * - Redirecting unauthenticated users to sign-in
 * - Rendering children only when user is authenticated
 * - Preserving intended destination for post-auth redirect
 * - Working with Next.js middleware for server-side protection
 */
export function AuthGate({ children, redirectTo = '/auth/sign-in', fallback }: AuthGateProps) {
  const { isAuthenticated, loading, initialized } = useAuth()
  const router = useRouter()

  useEffect(() => {
    // Only redirect after auth is initialized and user is not authenticated
    if (initialized && !loading && !isAuthenticated) {
      log.info('AuthGate', 'User not authenticated, redirecting to sign-in', {
        redirectTo,
      })

      // Get current path for redirect after sign-in
      const currentPath = typeof window !== 'undefined' ? window.location.pathname : '/'
      const signInUrl = `${redirectTo}?redirectTo=${encodeURIComponent(currentPath)}`

      // Use replace to avoid adding to navigation stack
      router.replace(signInUrl as any)
    }
  }, [initialized, loading, isAuthenticated, redirectTo, router])

  // Show loading state while auth is initializing
  if (!initialized || loading) {
    log.info('AuthGate', 'Auth initializing, showing loading state', {
      initialized,
      loading,
    })

    if (fallback) {
      return <>{fallback}</>
    }

    return (
      <YStack
        flex={1}
        justifyContent="center"
        alignItems="center"
        gap="$4"
        backgroundColor="$background"
        minHeight="100%"
      >
        <Spinner size="large" />
        <H3 color="$color10">Loading...</H3>
      </YStack>
    )
  }

  // Don't render children if not authenticated (redirect will happen)
  if (!isAuthenticated) {
    log.info('AuthGate', 'User not authenticated, not rendering children')
    return null
  }

  // User is authenticated, render protected content
  log.info('AuthGate', 'User authenticated, rendering protected content')
  return <>{children}</>
}
