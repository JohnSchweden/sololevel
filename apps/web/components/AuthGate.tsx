import { useAuth } from '@my/app/hooks/useAuth'
import { log } from '@my/logging'
import { H3, Spinner, YStack } from '@my/ui'
import { useRouter } from 'expo-router'
import React, { useEffect, useRef } from 'react'

export interface AuthGateProps {
  children: React.ReactNode
  redirectTo?: string
  fallback?: React.ReactNode
}

/**
 * AuthGate - Route protection component for Expo Router web app
 *
 * Protects child routes by:
 * - Showing loading state while auth is initializing
 * - Redirecting unauthenticated users to sign-in
 * - Rendering children only when user is authenticated
 * - Preserving intended destination for post-auth redirect
 * - Client-side route protection for Expo Router
 */
export function AuthGate({ children, redirectTo = '/auth/sign-in', fallback }: AuthGateProps) {
  const { isAuthenticated, loading, initialized } = useAuth()
  const router = useRouter()

  // Log deduplication ref - must be called unconditionally at component level
  const lastLoggedKeyRef = useRef<string>('')

  // Log deduplication helper to prevent spam of the same transition
  const logOnce = (key: string, fn: () => void) => {
    if (lastLoggedKeyRef.current !== key) {
      lastLoggedKeyRef.current = key
      fn()
    }
  }

  useEffect(() => {
    // Only redirect after auth is initialized and user is not authenticated
    if (initialized && !loading && !isAuthenticated) {
      log.info('AuthGate', 'User not authenticated, redirecting to sign-in', {
        redirectTo,
      })

      // Get current path for redirect after sign-in
      const currentPath = typeof window !== 'undefined' ? window.location.pathname : '/'
      const signInUrl = `${redirectTo}?redirectTo=${encodeURIComponent(currentPath)}`

      // Use replace to avoid adding to navigation stack
      router.replace(signInUrl as any)
    }
  }, [initialized, loading, isAuthenticated, redirectTo, router])

  // Show loading state while auth is initializing
  if (!initialized || loading) {
    if (typeof __DEV__ !== 'undefined' && __DEV__) {
      log.debug('AuthGate', 'Auth initializing, showing loading state', {
        initialized,
        loading,
      })
    }

    if (fallback) {
      return <>{fallback}</>
    }

    return (
      <YStack
        flex={1}
        justifyContent="center"
        alignItems="center"
        gap="$4"
        backgroundColor="$background"
        minHeight="100%"
      >
        {/* @ts-ignore - TS union type complexity limit */}
        <Spinner size="large" />
        <H3 color="$color10">Loading...</H3>
      </YStack>
    )
  }

  // Don't render children if not authenticated (redirect will happen)
  if (!isAuthenticated) {
    logOnce('not-authenticated', () =>
      log.info('AuthGate', 'User not authenticated, not rendering children')
    )
    return null
  }

  // User is authenticated, render protected content
  logOnce('authenticated', () =>
    log.info('AuthGate', 'User authenticated, rendering protected content')
  )
  return <>{children}</>
}
