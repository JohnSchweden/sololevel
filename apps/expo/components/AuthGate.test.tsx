import { useAuth } from '@my/app/hooks/useAuth'
import { useRouter } from 'expo-router'
import React from 'react'
import TestRenderer from 'react-test-renderer'
import { AuthGate } from './AuthGate'

jest.mock('@my/app/hooks/useAuth', () => ({
  useAuth: jest.fn(),
}))

const mockLogInfo = jest.fn()
const mockLogWarn = jest.fn()
jest.mock('@my/logging', () => ({
  log: {
    info: (...args: unknown[]) => mockLogInfo(...args),
    warn: (...args: unknown[]) => mockLogWarn(...args),
    debug: jest.fn(),
  },
}))

describe('AuthGate', () => {
  const mockRouter = {
    push: jest.fn(),
    replace: jest.fn(),
    back: jest.fn(),
  }

  beforeEach(() => {
    jest.clearAllMocks()
    jest.mocked(useRouter).mockReturnValue(mockRouter as any)
    jest.mocked(useAuth).mockReset()
  })

  it('renders children when user is authenticated', () => {
    jest.mocked(useAuth).mockReturnValue({
      user: { id: 'test-user-id' } as any,
      session: { access_token: 'test-token' } as any,
      loading: false,
      initialized: true,
      userId: 'test-user-id',
      isAuthenticated: true,
      signIn: jest.fn(),
      signOut: jest.fn(),
      initialize: jest.fn(),
    })

    let component: TestRenderer.ReactTestRenderer
    TestRenderer.act(() => {
      component = TestRenderer.create(
        <AuthGate>
          <React.Fragment>Protected Content</React.Fragment>
        </AuthGate>
      )
    })

    const tree = component!.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('shows loading when auth is not initialized', () => {
    jest.mocked(useAuth).mockReturnValue({
      user: null,
      session: null,
      loading: true,
      initialized: false,
      userId: null,
      isAuthenticated: false,
      signIn: jest.fn(),
      signOut: jest.fn(),
      initialize: jest.fn(),
    })

    let component: TestRenderer.ReactTestRenderer
    TestRenderer.act(() => {
      component = TestRenderer.create(
        <AuthGate>
          <React.Fragment>Protected Content</React.Fragment>
        </AuthGate>
      )
    })

    const tree = component!.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('redirects to sign-in when user is not authenticated', () => {
    jest.mocked(useAuth).mockReturnValue({
      user: null,
      session: null,
      loading: false,
      initialized: true,
      userId: null,
      isAuthenticated: false,
      signIn: jest.fn(),
      signOut: jest.fn(),
      initialize: jest.fn(),
    })

    TestRenderer.act(() => {
      TestRenderer.create(
        <AuthGate>
          <React.Fragment>Protected Content</React.Fragment>
        </AuthGate>
      )
    })

    expect(mockRouter.replace).toHaveBeenCalledWith('/auth/sign-in')
  })

  it('preserves intended destination for post-auth redirect', () => {
    jest.mocked(useAuth).mockReturnValue({
      user: null,
      session: null,
      loading: false,
      initialized: true,
      userId: null,
      isAuthenticated: false,
      signIn: jest.fn(),
      signOut: jest.fn(),
      initialize: jest.fn(),
    })

    TestRenderer.act(() => {
      TestRenderer.create(
        <AuthGate redirectTo="/custom-redirect">
          <React.Fragment>Protected Content</React.Fragment>
        </AuthGate>
      )
    })

    expect(mockRouter.replace).toHaveBeenCalledWith('/custom-redirect')
  })
})
