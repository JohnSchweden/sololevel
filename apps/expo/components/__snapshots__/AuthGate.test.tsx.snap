// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthGate renders children when user is authenticated 1`] = `"Protected Content"`;

exports[`AuthGate shows loading when auth is not initialized 1`] = `
[
  "Loading spinner",
  "Loading...",
]
`;
